### DISCUSSION TITLE
Error with Pairs Plotting when loading results

### DISCUSSION THREAD

**Discussion by sftorres405** [quaquel/EMAworkbench#273] [Q&A]
Kind greetings,

I have been playing with the lake problem example and encountered a rather annoying bug. When using the pairs_plotting.pairs_scatter function on results recently generated, it works as expected. However, if I save the results, load them, and then try to plot them I get the following error:

TypeError                                 Traceback (most recent call last)
[<ipython-input-39-376a1cb87818>](https://localhost:8080/#) in <cell line: 4>()
      2 from ema_workbench.analysis import pairs_plotting
      3 
----> 4 fig, axes = pairs_plotting.pairs_scatter(experiments, outcomes, group_by="policy", legend=False)
      5 fig.set_size_inches(8, 8)
      6 plt.show()

3 frames
[/usr/local/lib/python3.10/dist-packages/ema_workbench/analysis/pairs_plotting.py](https://localhost:8080/#) in pairs_scatter(experiments, outcomes, outcomes_to_show, group_by, grouping_specifiers, ylabels, legend, point_in_time, filter_scalar, **kwargs)
    443     _logger.debug("generating pairwise scatter plot")
    444 
--> 445     prepared_data = prepare_pairs_data(
    446         experiments,
    447         outcomes,

[/usr/local/lib/python3.10/dist-packages/ema_workbench/analysis/plotting_util.py](https://localhost:8080/#) in prepare_pairs_data(experiments, outcomes, outcomes_to_show, group_by, grouping_specifiers, point_in_time, filter_scalar)
    675         raise EMAError("for pair wise plotting, more than one outcome needs to be provided")
    676 
--> 677     experiments, outcomes, outcomes_to_show, time, grouping_labels = prepare_data(
    678         experiments, None, outcomes, outcomes_to_show, group_by, grouping_specifiers, filter_scalar
    679     )

[/usr/local/lib/python3.10/dist-packages/ema_workbench/analysis/plotting_util.py](https://localhost:8080/#) in prepare_data(experiments, experiments_to_show, outcomes, outcomes_to_show, group_by, grouping_specifiers, filter_scalar)
    762                     grouping_specifiers = set(column_to_group_by)
    763                 else:
--> 764                     grouping_specifiers = make_continuous_grouping_specifiers(
    765                         column_to_group_by, grouping_specifiers
    766                     )

[/usr/local/lib/python3.10/dist-packages/ema_workbench/analysis/plotting_util.py](https://localhost:8080/#) in make_continuous_grouping_specifiers(array, nr_of_groups)
    644     minimum = np.min(array)
    645     maximum = np.max(array)
--> 646     step = (maximum - minimum) / nr_of_groups
    647     a = [(minimum + step * x, minimum + step * (x + 1)) for x in range(nr_of_groups)]
    648     assert a[0][0] == minimum

TypeError: unsupported operand type(s) for /: 'int' and 'NoneType'

Any help on how to solve this issue would be appreciated.

**Comment by quaquel** [comment]
Can you share the code you used for generating the results and the version number of the ema_workbench? I tried replicating the error and have failed to do so with the latest code from [master](/quaquel/EMAworkbench/tree/master).

**Reply by sftorres405** [reply]
The version of ema is 2.4.0 installed from pip

_This is the code I used to generate the scenarios_:

from ema_workbench import (perform_experiments, ema_logging, save_results, load_results, MultiprocessingEvaluator)
from ema_workbench.em_framework import samplers, util

ema_logging.log_to_stderr(ema_logging.INFO)

n_scenarios = 100
n_policies = 4
fn = F'/{n_scenarios} scenarios {n_policies} policies.tar.gz'
counter = util.Counter()
policies = samplers.sample_levers(lake_model, n_policies, 
                                      sampler=samplers.MonteCarloSampler())
with MultiprocessingEvaluator(lake_model) as evaluator:
        results = evaluator.perform_experiments(n_scenarios, policies)
save_results(results, fn)

#results = load_results(fn)

_To generate the pair plots:_
experiments, outcomes = results
from ema_workbench.analysis import pairs_plotting
fig, axes = pairs_plotting.pairs_scatter(experiments, outcomes, group_by="policy", legend=False)
fig.set_size_inches(8, 8)
plt.show()

If I keep the load_results line commented it works perfectly. If I load the results I get the message that they were loaded successfully, but when plotting I get the error from the original post.

**Reply by quaquel** [reply]
There is no need to sample the policies separately. You can just do the following:

```python
n_scenarios = 100
n_policies = 4
fn = f'/{n_scenarios} scenarios {n_policies} policies.tar.gz'

with MultiprocessingEvaluator(lake_model) as evaluator:
    results = evaluator.perform_experiments(n_scenarios, n_policies, lever_sampling=samplers.MonteCarloSampler())
save_results(results, fn)

```

I'll try to find some time to have another look and see if anything relevant changed between 2.4 and master over the last 6 weeks that might explain the error and suggest a fix.

**Comment by EwoutH** [comment]
You're using `group_by="policy"`, and apparently `nr_of_groups` is `None` instead of an `int`, implicating that you don't have any groups. Do you have policies defined? And is there a policy column you your dataframe?

**Reply by quaquel** [reply]
There is always a policy column, but the value might be None. I'll test that later today.
